### ISSUE TITLE
The application state is not preserved when Wiredash is opened

### ISSUE THREAD

**Issue by danielkauffmann** [wiredashio/wiredash-sdk#289]
**Describe the bug**
Hello everyone, I'd like to first congratulate the team for the product you've created. I've been using Wiredash for about 2 years now, and from the very first day, it captured me both visually and in how it simplifies the feedback process by presenting complex features in a user-friendly manner that completes the flow. The work this team has done is incredible.

This bug arose from a recent discussion I had on #287, and based on my understanding of the product, the following use case is a bug. Even though it's a simple issue, what happens is that **the application state is not preserved when Wiredash is opened.**

The expected behavior was for the states of the Widgets below the tree where the user would provide feedback to be maintained. In my opinion, this shouldn't happen, and I'm here to understand if there was a mistake in my implementation or if this behavior was intentional.

[You can access the repository here](https://github.com/danielkauffmann/wiredash-preserve-state) for reproducing the bug. If this behavior is indeed expected, I'd like to understand the reason behind this choice.

**Stacktrace**
Not applicable.

**Wiredash SDK Info**
_Which SDK are you using?_ 1.7.2
_On which platform are you seeing the crash?_ iOS 16.4

**Flutter SDK**
_Which Flutter version are you using? Run `flutter doctor -v`_
<details>
<summary>Output</summary>

```
$ flutter doctor -v

[✓] Flutter (Channel stable, 3.10.6, on macOS 13.5 22G74 darwin-arm64, locale en-BR)
    • Flutter version 3.10.6 on channel stable at /Users/daniel/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f468f3366c (4 weeks ago), 2023-07-12 15:19:05 -0700
    • Engine revision cdbeda788a
    • Dart version 3.0.6
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/daniel/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/daniel/Library/Android/sdk
    • Java binary at: /Users/daniel/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/223.8836.35.2231.10320515/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio
    • Android Studio at /Users/daniel/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/223.8836.35.2231.10271316/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/daniel/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/223.8836.35.2231.10320515/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.81.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.70.0

[✓] Connected device (3 available)
    • iPhone 14 Pro Max (mobile) • 6F06BF58-187F-414E-9FFF-48DE0BE52759 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 13.5 22G74 darwin-arm64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 114.0.5735.106

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

**Screenshots**
<details>
<summary>Initial state</summary>
<img src='https://github.com/wiredashio/wiredash-sdk/assets/62577994/bc9302b5-d125-42f0-8923-ea4b5dd429dc' width='300'/>
</details>

<details>
<summary>Once Wiredash is opened</summary>
<img src='https://github.com/wiredashio/wiredash-sdk/assets/62577994/d44554b7-5168-4b79-a9eb-d3655c1878c6' width='300'/>
</details>

**Comment by passsy** [comment]
Thanks for the detailed bug report, only this way I'm able to help you

First of all, Wiredash preserves the state of the child widget. There is even a [test](https://github.com/wiredashio/wiredash-sdk/blob/5ac0b9dd2658fbafcab3bf822730f7168a729fb6/test/wiredash_widget_test.dart#L136) verifying it.

When Wiredash opens, new widgets are inserted into the widget tree above your app. The state of the app is saved though, with a global key. 
Two things happen to your app widget:
1. `didChangeDependencies` is called, informing your app that the parent widget structure has changed and your widget has been moved in the widget tree
2. `build` is called to rerender your widget.

<details>
<summary>Sample code</summary>

```dart
import 'package:flutter/material.dart';
import 'package:wiredash/wiredash.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return Wiredash(
      projectId: 'YOUR_PROJECT_ID',
      secret: 'YOUR_SECRET',
      child: MaterialApp(
        title: 'Wiredash preserve state demo',
        theme: ThemeData(
          colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
          useMaterial3: true,
        ),
        home: const MyHomePage(),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  void initState() {
    super.initState();
    print('initState');
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    print('didChangeDependencies');
  }

  @override
  void dispose() {
    super.dispose();
    print('dispose');
  }

  @override
  Widget build(BuildContext context) {
    print('build');
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: const Text('Wiredash preserve state demo'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FutureBuilder(
              future: Future.delayed(const Duration(seconds: 1)),
              builder: (context, snapshot) {
                if (snapshot.connectionState == ConnectionState.done) {
                  return const Text('FutureBuilder finished!');
                }

                return const CircularProgressIndicator();
              },
            ),
            ElevatedButton(
              onPressed: () {
                print('show wiredash');
                Wiredash.of(context).show();
              },
              child: const Text('Show Wiredash'),
            )
          ],
        ),
      ),
    );
  }
}
```

</details>

I updated your sample code and added logging. This is the output:

```dart
flutter: initState
flutter: didChangeDependencies
flutter: build
flutter: build
flutter: show wiredash
flutter: didChangeDependencies
flutter: build
flutter: didChangeDependencies
flutter: build
```

Note: `initState` is only called once, `dispose` is not called at all. Both confirm that Wiredash does indeed preserve the state of the widget correctly.

--- 

The issue you're having in your app is one of the most common problems in Flutter. The misunderstanding of `FutureBuilder`. The problematic part is this line: 

```dart
            FutureBuilder(
              future: Future.delayed(const Duration(seconds: 1)),
```

This creates a new future, every time the `build` method is called. Which is not what you want! You want the future to be created only once. You also can't stop Flutter from rebuilding your widget. This is just how it works, you have to handle the case that `build` is called multiple times.

This problem was covered explicitly in this video: https://www.youtube.com/watch?v=zEdw_1B7JHY

The solution is easy: create the `Future` in `initState` like this:

```dart
class _MyHomePageState extends State<MyHomePage> {
  late Future<void> _future;
  @override
  void initState() {
    super.initState();
    _future = Future.delayed(const Duration(seconds: 1));
  }


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: const Text('Wiredash preserve state demo'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FutureBuilder(
              future: _future,
              builder: (context, snapshot) {
                if (snapshot.connectionState == ConnectionState.done) {
                  return const Text('FutureBuilder finished!');
                }

                return const CircularProgressIndicator();
              },
            ),
          ],
        ),
      ),
    );
  }
}

```

<details>
<summary>Full code</summary>


```dart
import 'package:flutter/material.dart';
import 'package:wiredash/wiredash.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return Wiredash(
      projectId: 'YOUR_PROJECT_ID',
      secret: 'YOUR_SECRET',
      child: MaterialApp(
        title: 'Wiredash preserve state demo',
        theme: ThemeData(
          colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
          useMaterial3: true,
        ),
        home: const MyHomePage(),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late Future<void> _future;
  @override
  void initState() {
    super.initState();
    print('initState');
    _future = Future.delayed(const Duration(seconds: 1));
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    print('didChangeDependencies');
  }

  @override
  void dispose() {
    super.dispose();
    print('dispose');
  }

  @override
  Widget build(BuildContext context) {
    print('build');
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: const Text('Wiredash preserve state demo'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FutureBuilder(
              future: _future,
              builder: (context, snapshot) {
                if (snapshot.connectionState == ConnectionState.done) {
                  return const Text('FutureBuilder finished!');
                }

                return const CircularProgressIndicator();
              },
            ),
            ElevatedButton(
              onPressed: () {
                print('show wiredash');
                Wiredash.of(context).show();
              },
              child: const Text('Show Wiredash'),
            )
          ],
        ),
      ),
    );
  }
}

```

</details>

Note: This not only fixes the problem you're having when opening Wiredash. It also fixes Hot Reload which was not working as expected in the sample you provided.

**Comment by danielkauffmann** [comment]
Thank you very much, @passsy, for your detailed response! It all made sense, and I understood where I went wrong in my previous example.

One of the motivations behind this issue (also related to #287) was according to the router used in my application. Going deeper and testing other routers like [GoRouter](https://pub.dev/packages/go_router), I found out that the state preservation issue occurs in [AutoRoute](https://pub.dev/packages/auto_route) itself when I use it with Wiredash. I will open an issue in its repository.

I appreciate your patience and feedback!!
